File: algorithms/002.go
package main

import "fmt"

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	n1 := l1
	n2 := l2
	var leading_node *ListNode
	var last_node *ListNode
	moving_up := 0
	for n1 != nil || n2 != nil {
		v1, v2, sum := 0, 0, 0
		if n1 != nil {
			v1 = n1.Val
			n1 = n1.Next
		}
		if n2 != nil {
			v2 = n2.Val
			n2 = n2.Next
		}
		sum = v1 + v2 + moving_up
		moving_up = sum / 10
		n := &ListNode{Val: sum % 10}
		if leading_node != nil {
			last_node.Next = n
			last_node = n
		} else {
			leading_node = n
			last_node = n
		}
	}
	if moving_up > 0 {
		n := &ListNode{Val: moving_up}
		last_node.Next = n
	}
	return leading_node
}

func main() {
	// l1: 60
	l1 := &ListNode{Val: 0}
	l1.Next = &ListNode{Val: 6}
	// l2: 972
	l2 := &ListNode{Val: 2}
	l2.Next = &ListNode{Val: 7}
	l2.Next.Next = &ListNode{Val: 9}
	got := addTwoNumbers(l1, l2)
	if got.Val != 2 || got.Next.Val != 3 || got.Next.Next.Val != 0 || got.Next.Next.Next.Val != 1 {
		fmt.Printf("%d\n", got.Val)
		fmt.Printf("%d\n", got.Next.Val)
		fmt.Printf("%d\n", got.Next.Next.Val)
		fmt.Printf("%d\n", got.Next.Next.Next.Val)
	}
}

File: algorithms/002.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):

    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        l = []
        n1 = l1
        n2 = l2
        moving_up = 0
        while n1 is not None or n2 is not None:
            v1 = 0
            if n1 is not None:
                v1 = n1.val
                n1 = n1.next
            v2 = 0
            if n2 is not None:
                v2 = n2.val
                n2 = n2.next
            val = v1 + v2 + moving_up
            moving_up = int(val / 10)
            l.append(val % 10)
        if moving_up != 0:
            l.append(moving_up)
        return self.getLinkedListNodes(l)

    def getLinkedListNodes(self, l):
        """
        :type l: list[int]
        """
        leading_node = ListNode(l[-1])
        for n in reversed(l[:-1]):
            node = ListNode(n)
            node.next = leading_node
            leading_node = node
        return leading_node


if __name__ == '__main__':
    sol = Solution()
    l1 = sol.getLinkedListNodes([2, 4, 3])
    l2 = sol.getLinkedListNodes([5, 6, 4])
    ans = sol.addTwoNumbers(l1, l2)
    assert ans.val == 7 and ans.next.val == 0 and ans.next.next.val == 8
    l1 = sol.getLinkedListNodes([2, 5, 8])
    l2 = sol.getLinkedListNodes([0, 8, 3])
    ans = sol.addTwoNumbers(l1, l2)
    assert ans.val == 2 and ans.next.val == 3 and ans.next.next.val == 2 \
        and ans.next.next.next.val == 1
    l1 = sol.getLinkedListNodes([1, 8])
    l2 = sol.getLinkedListNodes([0])
    ans = sol.addTwoNumbers(l1, l2)
    assert ans.val == 1 and ans.next.val == 8
